### DISCUSSION TITLE
What is the best g2p for japanese TTS?

### DISCUSSION THREAD

**Discussion by aidiary** [espnet/espnet#5007] [Q&A] [answered]
There are around 5 different G2P (grapheme-to-phoneme) conversion methods in Japanese, as listed [here](https://github.com/espnet/espnet/blob/master/espnet2/text/phoneme_tokenizer.py#L18-L22.). Do you happen to know which method is considered the best?

I tested VITS TTS on my dataset using two methods, `pyopenjtalk_accent_with_pause` and `pyopenjtalk_prosody`, and found both to be effective.

However, during inference, `pyopenjtalk_prosody` sometimes missed some phonemes, possibly due to a duration prediction issue.

Initially, I thought `pyopenjtalk_accent_with_pause` would be the best option for VITS, but I realized it was unable to differentiate between 。 and ？ as their token sequences were the same.

```
% from espnet2.text.phoneme_tokenizer import pyopenjtalk_g2p_accent_with_pause
% declarative = pyopenjtalk_g2p_accent_with_pause("これはテストですか。")
% interrogative = pyopenjtalk_g2p_accent_with_pause("これはテストですか？")
% print(declarative == interrogative)
True
```

Do you have any advice for me?
Thank you.

**Comment by kan-bayashi** [comment, marked as answer]
Hi @aidiary, thank you for raising interesting discussion.

In my experiences, `pyopenjtalk_prosody` and `pyopenjtalk_accent_with_pause` are almost the same in terms of quality.
But in terms of easiness of manual modification, I think `pyopenjtalk_prosody` is the best due to its simple and understandable representations.

> However, during inference, pyopenjtalk_prosody sometimes missed some phonemes, possibly due to a duration prediction issue.

This is interesting. I didn't notice such a behavior.
Does it work without any issue when using `pyopenjtalk_accent_with_pause`?

> I realized it was unable to differentiate between 。 and ？ as their token sequences were the same.

This is a good point. Maybe we can extend `pyopenjtalk_accent_with_pause` to keep the such symbols.
The code itself is very simple, so you can easily extend it.
https://github.com/espnet/espnet/blob/611a2916a736dc07112ecf143faadadf9ed90e1e/espnet2/text/phoneme_tokenizer.py#L84-L93

P.S.
When I was an undergraduated student, I studied a lot from your blog. 
Thank you so much for sharing your knowledge :)

**Reply by aidiary** [reply]
Thank you for your advice.

I tried to compare `pyopenjtalk_accent_with_pause` and `pyopenjtalk_prosody` by fine-tuning in my corpus, but the problem occurred only in `pyopenjtalk_prosody`.

I also tried [this pre-trained model](https://huggingface.co/espnet/kan-bayashi_jsut_tts_train_full_band_vits_raw_phn_jaconv_pyopenjtalk_p-truncated-66d5fc) that was trained with JSUT, but did not encounter such an issue. Therefore, the problem may depend on the corpus or hyperparameters. I will continue to investigate further.

[This blog post](https://engineering.linecorp.com/ja/blog/end-to-end-tts) also mentioned the problem of phoneme duration shortening in VITS. Although I am unsure which G2P the author is using, it seems to be the same problem.

I will further investigate the issue with `pyopenjtalk_prosody`, and if I cannot solve it, I will attempt to improve `pyopenjtalk_accent_with_pause`.

P.S. Thank you for reading my blog. We have started to use espnet extensively in my research and development. Thank you for creating such a great OSS project!
